terraform-provider-lxd_v2.4.0 crashed with nil pointer dereference

### Description

When I run `terraform init` while using LXD Terraform provider, it panicked with nil pointer dereference.
I realized that I forgot to install Juju. However, it can return a meaningful error instead of crashing.
Then, I installed Juju and run `terraform init` again. It keeps crashing.

### Urgency

Casually reporting

### Terraform Juju Provider version

2.4.0

### Terraform version

1.10.5

### Juju version

3.6.2

### Terraform Configuration(s)

```terraform

```

### Reproduce / Test

```shell
Without installing Juju, run `terrafom init` and `terraform apply`.
```

### Debug/Panic Output

```shell
gatici@gaticipc:~/charmed5g/charmed-aether-sd-core/terraform$ terraform init
terraform apply -auto-approve
Initializing the backend...
Initializing provider plugins...
- Reusing previous version of terraform-lxd/lxd from the dependency lock file
- Reusing previous version of hashicorp/tls from the dependency lock file
- Using previously-installed terraform-lxd/lxd v2.4.0
- Using previously-installed hashicorp/tls v4.0.6

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
tls_private_key.juju-key: Refreshing state... [id=70f1efdf5c389e3e9a8335b69ad954aa210fd112]
lxd_storage_pool.sdcore-pool: Refreshing state... [name=sdcore-pool]
lxd_network.sdcore-core: Refreshing state... [name=sdcore-core]
lxd_network.sdcore-access: Refreshing state... [name=sdcore-access]
lxd_network.sdcore-mgmt: Refreshing state... [name=sdcore-mgmt]
lxd_network.sdcore-ran: Refreshing state... [name=sdcore-ran]
lxd_instance.control-plane: Refreshing state... [name=control-plane]
lxd_instance_file.control-plane-privkey: Refreshing state...
lxd_instance_file.control-plane-pubkey: Refreshing state...

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform planned the following actions, but then encountered a problem:

  # lxd_instance.control-plane is tainted, so must be replaced
-/+ resource "lxd_instance" "control-plane" {
      + execs            = {
          + "00-wait-for-boot" = {
              + command       = [
                  + "systemctl",
                  + "is-system-running",
                  + "--wait",
                  + "--quiet",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = false
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "on_start"
              + working_dir   = ""
            },
          + "01-install-microk8s" = {
              + command       = [
                  + "snap",
                  + "install",
                  + "microk8s",
                  + "--channel=1.31-strict/stable",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = true
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "02-wait-for-microk8s" = {
              + command       = [
                  + "microk8s",
                  + "status",
                  + "--wait",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = true
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "03-microk8s-disable-default-dns" = {
              + command       = [
                  + "microk8s",
                  + "disable",
                  + "dns",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = false
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "04-microk8s-enable-custom-dns" = {
              + command       = [
                  + "microk8s",
                  + "enable",
                  + "dns:10.201.0.1",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = false
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "05-microk8s-enable-hostpath-storage" = {
              + command       = [
                  + "microk8s",
                  + "enable",
                  + "hostpath-storage",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = false
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "06-microk8s-enable-metallb" = {
              + command       = [
                  + "microk8s",
                  + "enable",
                  + "metallb:10.201.0.52-10.201.0.53",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = false
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "07-add-ubuntu-user-to-snap_microk8s-group" = {
              + command       = [
                  + "usermod",
                  + "-a",
                  + "-G",
                  + "snap_microk8s",
                  + "ubuntu",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = true
              + record_output = false
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
          + "08-get-microk8s-config" = {
              + command       = [
                  + "microk8s.config",
                ]
              + enabled       = true
              + environment   = {}
              + exit_code     = (known after apply)
              + fail_on_error = true
              + record_output = true
              + run_count     = (known after apply)
              + stderr        = (known after apply)
              + stdout        = (known after apply)
              + trigger       = "once"
              + working_dir   = ""
            },
        }
      + image            = "ubuntu:24.04"
      ~ interfaces       = {
          - "eth0" = {
              - ips   = [
                  - {
                      - address = "10.201.0.101" -> null
                      - family  = "inet" -> null
                      - scope   = "global" -> null
                    },
                  - {
                      - address = "fe80::216:3eff:fe51:9f1a" -> null
                      - family  = "inet6" -> null
                      - scope   = "link" -> null
                    },
                ] -> null
              - name  = "enp5s0" -> null
              - state = "up" -> null
              - type  = "broadcast" -> null
            },
        } -> (known after apply)
      ~ ipv4_address     = "10.201.0.101" -> (known after apply)
      + ipv6_address     = (known after apply)
      ~ mac_address      = "00:16:3e:51:9f:1a" -> (known after apply)
        name             = "control-plane"
      ~ status           = "Running" -> (known after apply)
      ~ target           = null -> (known after apply)
      + timeouts         = {
          + create = "10m"
          + delete = "10m"
          + read   = "10m"
          + update = "10m"
        }
        # (8 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ValidateResourceConfig call. The plugin logs may contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-lxd_v2.4.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc60554]

goroutine 148 [running]:
github.com/terraform-lxd/terraform-provider-lxd/internal/instance.InstanceFileResource.Read({0x0?}, {0x1056918, 0xc0007a0240}, {{{{0x105cc90, 0xc0007a1500}, {0xd8f360, 0xc0007a1380}}, {0x105f418, 0xc0003a18b0}}, 0xc00038c050, ...}, ...)
        github.com/terraform-lxd/terraform-provider-lxd/internal/instance/resource_instance_file.go:274 +0x634
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ReadResource(0xc00025e1e0, {0x1056918, 0xc0007a0240}, 0xc0007a02a0, 0xc000565618)
        github.com/hashicorp/terraform-plugin-framework@v1.12.0/internal/fwserver/server_readresource.go:117 +0x84e
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ReadResource(0xc00025e1e0, {0x1056918?, 0xc0007a0150?}, 0xc0007a6080)
        github.com/hashicorp/terraform-plugin-framework@v1.12.0/internal/proto6server/server_readresource.go:55 +0x38e
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ReadResource(0xc0002535e0, {0x1056918?, 0xc000888090?}, 0xc0002dc310)
        github.com/hashicorp/terraform-plugin-go@v0.24.0/tfprotov6/tf6server/server.go:784 +0x309
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ReadResource_Handler({0xeadbe0, 0xc0002535e0}, {0x1056918, 0xc000888090}, 0xc00079a000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.24.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:509 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001d6e00, {0x1056918, 0xc000888000}, {0x105dea0, 0xc000476000}, 0xc0007b4000, 0xc000362e10, 0x16a5368, 0x0)
        google.golang.org/grpc@v1.66.2/server.go:1394 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc0001d6e00, {0x105dea0, 0xc000476000}, 0xc0007b4000)
        google.golang.org/grpc@v1.66.2/server.go:1805 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.2/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 35
        google.golang.org/grpc@v1.66.2/server.go:1040 +0x125

Error: The terraform-provider-lxd_v2.4.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Notes & References

_No response_

--- a/internal/instance/resource_instance_file.go
+++ b/internal/instance/resource_instance_file.go
@@ -263,9 +263,16 @@ func (r InstanceFileResource) Read(ctx context.Context, req resource.ReadRequest
 		return
 	}
 
-	// Fetch file
+	// Fetch an existing file.
 	_, file, err := server.GetInstanceFile(instanceName, targetPath)
 	if err != nil {
+		if errors.IsNotFoundError(err) {
+			// If file is not found, remove it from the Terraform state
+			// to ensure it is recreated.
+			resp.State.RemoveResource(ctx)
+			return
+		}
+
 		resp.Diagnostics.AddError(fmt.Sprintf("Failed to retrieve file %q from instance %q", targetPath, instanceName), err.Error())
 		return
 	}

--- a/internal/instance/resource_instance_file.go
+++ b/internal/instance/resource_instance_file.go
@@ -267,6 +267,7 @@ func (r InstanceFileResource) Read(ctx context.Context, req resource.ReadRequest
 	_, file, err := server.GetInstanceFile(instanceName, targetPath)
 	if err != nil {
 		resp.Diagnostics.AddError(fmt.Sprintf("Failed to retrieve file %q from instance %q", targetPath, instanceName), err.Error())
+		return
 	}
 
 	state.Instance = types.StringValue(instanceName)

--- a/internal/instance/resource_instance_file.go
+++ b/internal/instance/resource_instance_file.go
@@ -324,20 +324,9 @@ func (r InstanceFileResource) Delete(ctx context.Context, req resource.DeleteReq
 		return
 	}
 
-	file := common.InstanceFileModel{
-		Content:    state.Content,
-		SourcePath: state.SourcePath,
-		TargetPath: state.TargetPath,
-		UserID:     state.UserID,
-		GroupID:    state.GroupID,
-		Mode:       state.Mode,
-		CreateDirs: state.CreateDirs,
-		Append:     state.Append,
-	}
-
 	// Delete file.
-	err = common.InstanceFileUpload(server, instanceName, file)
-	if err != nil {
+	err = common.InstanceFileDelete(server, instanceName, state.TargetPath.ValueString())
+	if err != nil && !errors.IsNotFoundError(err) {
 		resp.Diagnostics.AddError(fmt.Sprintf("Failed to delete file %q from instance %q", targetFile, instanceName), err.Error())
 		return
 	}

--- a/internal/instance/resource_instance_file.go
+++ b/internal/instance/resource_instance_file.go
@@ -256,6 +256,9 @@ func (r InstanceFileResource) Read(ctx context.Context, req resource.ReadRequest
 	instance, _, err := server.GetInstance(instanceName)
 	if err != nil {
 		if errors.IsNotFoundError(err) {
+			// If instance is not found, file cannot exist. Remove it
+			// from the state to ensure it is created on next apply.
+			resp.State.RemoveResource(ctx)
 			return
 		}
 

